Removing the metadata instantiation from the EM

mikeSimonson opened the pull request:
The metadata instantiation is kept in the create method for BC for now.
A new static method is added to be able to pass the metadata to the EM.

## TODOs

 * [ ] Document BC Breakages
 * [ ] Add lazy collection implementation (populated as-we-go)
 * [ ] Ensure type aliases are correctly cached (create key copies in the metadata)
 * [ ] Ensure the `\\` `ltrim()` is gone

mikeSimonson committed:
Removing the metadata instantiation from the EM

The metadata instantiation is kept in the create method for BC for now.
A new static method is added to be able to pass the metadata to the EM.

mikeSimonson commented:
@Ocramius Is this what you  meant when we discussed at phpbenelux ?
Is the next step refactoring all the call to create to the new createWithMetadata function ?

Ocramius commented:
@mikeSimonson the idea was to make the metadata a `ClassMetadata[]` lazy collection (injectable) rather than a `ClassMetadataFactory` - is this a middle-ground before we go there?

mikeSimonson commented:
I just wanted to make sure that I was going in the right direction.
I should pull lazymap as a dep then I suppose ?

Ocramius commented:
@mikeSimonson I'd suggest making it an eager map for now, and later we fix it with `LazyMap`

mikeSimonson committed:
WIP

mikeSimonson commented:
Does the new commit goes in the right direction ?

Can I
    get rid of the [cache configuration in the EM constructor](https://github.com/doctrine/doctrine2/pull/7143/files#diff-cfa931dd1195fd8c5e36f3706b028dbaR153) ?
    change the behavior of [getAllMetadata](https://github.com/doctrine/doctrine2/pull/7143/files#diff-dfe17e5e229e8d404df56b3c4849d511R135) for simplicity ?
    remove [the initialize thingy in the ClassMetadataFactory](https://github.com/doctrine/doctrine2/pull/7143/files#diff-ecaa162f62bde19c758286c766911141R83) ?

How do I proceed to remove [the entity manager from the OnClassMetadataNotFoundEventArgs in the ClassMetadataFactory](https://github.com/mikeSimonson/doctrine2/blob/bc4133e3e4f530285168afacc95c591a4679eccc/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php#L97) and [LoadClassMetadataEventArgs](https://github.com/mikeSimonson/doctrine2/blob/bc4133e3e4f530285168afacc95c591a4679eccc/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php#L166) ?

Any other remarks / comment / direction ?
@Ocramius

mikeSimonson committed:
WIP EM tests pass

Ocramius commented on lib/Doctrine/ORM/EntityManager.php:
```diff
@@ -155,7 +155,13 @@ protected function __construct(Connection $conn, Configuration $config, EventMan
+    /**
+     * @return MetadataCollection
+     */
+    public function getMetadatas(): MetadataCollection
```
Would avoid exposing it here at all.

Ocramius commented on lib/Doctrine/ORM/EntityManager.php:
```diff
@@ -134,52 +128,47 @@
+     * @return MetadataCollection
      */
-    public function getConnection()
+    public function getMetadatas(): MetadataCollection
```
Avoid exposing this

Ocramius commented on lib/Doctrine/ORM/EntityManager.php:
```diff
@@ -273,7 +262,8 @@ public function rollback()
     public function getClassMetadata($className) : Mapping\ClassMetadata
     {
-        return $this->metadataFactory->getMetadataFor($className);
+        $className = StaticClassNameConverter::getRealClass($className);
```
This can likely be dealt with in a `MetadataCollection` decorator instead

Ocramius commented on lib/Doctrine/ORM/EntityManager.php:
```diff
@@ -134,52 +128,47 @@
+        $this->conn            = $conn;
+        $this->config          = $config;
+        $this->eventManager    = $eventManager;
+        $this->metadatas       = $metadataCollection;
```
`metadata` is irregular: the plural of `metadata` is `metadata`. A better way to name this is probably `mappings`

Ocramius commented on lib/Doctrine/ORM/EntityManager.php:
```diff
@@ -382,8 +372,13 @@ public function find($entityName, $id, $lockMode = null, $lockVersion = null)
+                try{
+                    $this->metadatas->get(StaticClassNameConverter::getClass($value));
+                    $id[$i] = $this->unitOfWork->getSingleIdentifierValue($value);
+                } catch (\Exception $e) {
```
Most likely wrong here

Ocramius commented on lib/Doctrine/ORM/EntityManager.php:
```diff
@@ -382,8 +372,13 @@ public function find($entityName, $id, $lockMode = null, $lockVersion = null)
-                $id[$i] = $this->unitOfWork->getSingleIdentifierValue($value);
+            if (is_object($value)) {
+                try{
+                    $this->metadatas->get(StaticClassNameConverter::getClass($value));
```
Same here: move `StaticClassNameConverter` into a detail of `MetadataCollection`

Ocramius commented on lib/Doctrine/ORM/EntityManager.php:
```diff
@@ -479,8 +474,13 @@ public function getReference($entityName, $id)
+                try{
+                    $this->metadatas->get(StaticClassNameConverter::getClass($value));
+                    $scalarId[$i] = $this->unitOfWork->getSingleIdentifierValue($value);
+                } catch (\Exception $e) {
```
Likely wrong here - why did you introduce these try-catches? I know you are still playing around with the patch: just wondering why this got added to make things work

Ocramius commented on lib/Doctrine/ORM/EntityManager.php:
```diff
@@ -541,8 +541,13 @@ public function getPartialReference($entityName, $id)
+                try{
+                    $this->metadatas->get(StaticClassNameConverter::getClass($value));
+                    $id[$i] = $this->unitOfWork->getSingleIdentifierValue($value);
+                } catch (\Exception $e) {
```
Likely wrong here - why did you introduce these try-catches? I know you are still playing around with the patch: just wondering why this got added to make things work

mikeSimonson replied:
just because for now get throws an Exception if the metadata is not found where it wasn't before.

Ocramius commented on lib/Doctrine/ORM/EntityManager.php:
```diff
@@ -829,7 +834,23 @@ public static function create($connection, Configuration $config, ?EventManager
-        return new EntityManager($connection, $config, $connection->getEventManager());
+        $metadataFactoryClassName = $config->getClassMetadataFactoryName();
+        /** @var \Doctrine\Common\Persistence\Mapping\ClassMetadataFactory $metadataFactory */
+        $metadataFactory          = new $metadataFactoryClassName($config, $connection, $eventManager);
```
:+1: this is a decent BC-compliant approach for now :)

Ocramius commented on lib/Doctrine/ORM/Mapping/MetadataCollection.php:
```diff
@@ -0,0 +1,43 @@
+        return $this->classMetadatas[$name];
+    }
+
+    public static function fromClassMetadatas(array $classMetadatas)
```
I suggest using `(ClassMetadata $firstClass, ClassMetadata ...$otherClasses)` here, which prevents instantiation if no metadata is available at all.

Ocramius commented on lib/Doctrine/ORM/Mapping/MetadataCollection.php:
```diff
@@ -0,0 +1,43 @@
+
+    public function get($name)
+    {
+        if(!isset($this->classMetadatas[$name])){
```
Add the class name normalisation here

Ocramius commented on lib/Doctrine/ORM/Mapping/MetadataCollection.php:
```diff
@@ -0,0 +1,43 @@
+
+    private function __construct(array $classMetadatas)
+    {
+        $metadataByClassName = array_reduce (
```
This is simpler with:

```php
array_combine(
    array_map(function (ClassMetadata $metadata) { return $metadata->getClassName(); }, $metadata),
    $metadata
);
```

Ocramius commented:
> Does the new commit goes in the right direction ?

:+1: 

 > Can I get rid of the [cache configuration in the EM constructor](https://github.com/doctrine/doctrine2/pull/7143/files#diff-cfa931dd1195fd8c5e36f3706b028dbaR153) ?

Most likely. Metadata caching can be moved to a `CachedMetadataCollection` or such

 > change the behavior of [getAllMetadata](https://github.com/doctrine/doctrine2/pull/7143/files#diff-dfe17e5e229e8d404df56b3c4849d511R135) for simplicity ?

That is a subtle BC break: please don't change that.

 > remove [the initialize thingy in the ClassMetadataFactory](https://github.com/doctrine/doctrine2/pull/7143/files#diff-ecaa162f62bde19c758286c766911141R83) ?

Yes: since metadata now has a lifecycle that is completely separate from the ORM itself, removal is fine here :+1: 

 > How do I proceed to remove [the entity manager from the OnClassMetadataNotFoundEventArgs in the ClassMetadataFactory](https://github.com/mikeSimonson/doctrine2/blob/bc4133e3e4f530285168afacc95c591a4679eccc/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php#L97) and [LoadClassMetadataEventArgs](https://github.com/mikeSimonson/doctrine2/blob/bc4133e3e4f530285168afacc95c591a4679eccc/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php#L166) ?

I suggest we completely remove these events, as the user can simply decorate the `MetadataCollection` in order to manipulate metadata information

 > Any other remarks / comment / direction ?

Provided above :-)

mikeSimonson committed:
Removing those event that won't be necessary anymore

mikeSimonson committed:
No need to manage the initalization lifecycle anymore

as it's completely separeted from the ORM lifecycle now, at least it's
the goal

mikeSimonson committed:
adding a property for the mapping driver

mikeSimonson committed:
removing subtle BC break

mikeSimonson committed:
Not exposing the metadatas from the entity manager

mikeSimonson committed:
Removing the left cache concern from the entity manager

mikeSimonson committed:
Moving the responsability of cleaning the class name to the collection

mikeSimonson committed:
simplifying the metadata collection

Making sure that it can't be instantiated with an empty array

mikeSimonson committed:
Correcting the property name for clarity

mikeSimonson committed:
Adding type hint and return type

mikeSimonson committed:
Making the Event manager required for now (debug easier)

mikeSimonson committed:
Moving the responsability of the canonicalizing the class name

mikeSimonson committed:
Fixing test WIP

mikeSimonson committed:
The StaticClassNameConverter is already called in the MetadataCollection

Majkl578 commented on lib/Doctrine/ORM/Mapping/MetadataCollection.php:
```diff
@@ -0,0 +1,41 @@
+        return $this->metadata[$name];
+    }
+
+    public static function fromClassMetadatas(ClassMetadata $firstClass, ClassMetadata ...$otherClasses)
```
Why is this a static instead of constructor directly?

Also the signature is ugly, let's just have `__construct(ClassMetadata ...$metadata)` with `assert(count($metadata) !== 0)` if it really MUST be non-empty.

Majkl578 commented on lib/Doctrine/ORM/EntityManager.php:
```diff
@@ -784,7 +773,7 @@ public function newHydrator($hydrationMode)
 
             default:
-                $class = $this->config->getCustomHydrationMode($hydrationMode);
+                $class = $this->config->getCustomHydrationMode((string) $hydrationMode);
```
This is most likely wrong, see https://github.com/doctrine/doctrine2/pull/7130#discussion_r175259739. It could be `int|string` currently.

mikeSimonson replied:
getCustomHydrationMode expect a string in it's function definition.

Majkl578 replied:
:thinking:

Majkl578 replied:
Changed this in my PR since it's really `int|string` now. Something to re-do later on probably.

mikeSimonson committed:
fix missed namespace migration

mikeSimonson committed:
WIP

mikeSimonson committed:
simplify the tests

mikeSimonson committed:
add typing

mikeSimonson committed:
extract the MetadataFactory from the basic entity persister

mikeSimonson committed:
Adapt entity manager and unit of work to the basic persister change

mikeSimonson commented:
@Ocramius My main question now is to know if passing the whole mapping collection to the BasicEntityPersister is the way to go ? (commit extract the MetadataFactory from the basic entity persister and the following one)

Ocramius commented on lib/Doctrine/ORM/EntityManager.php:
```diff
@@ -541,8 +525,13 @@ public function getPartialReference($entityName, $id)
-            if (is_object($value) && $this->metadataFactory->hasMetadataFor(StaticClassNameConverter::getClass($value))) {
-                $id[$i] = $this->unitOfWork->getSingleIdentifierValue($value);
+            if (is_object($value)) {
+                try{
```
Same: try-catch should be removed overall. We possibly need a better abstraction.

Ocramius commented on lib/Doctrine/ORM/Mapping/MetadataCollection.php:
```diff
@@ -0,0 +1,47 @@
+
+    public function get(string $name) : ClassMetadata
+    {
+        $name = StaticClassNameConverter::getRealClass($name);
```
Consider checking for existence before calling `StaticClassNameConverter`. Copying a metadata reference into a different key is also a good idea for performance reasons

Ocramius commented on lib/Doctrine/ORM/Mapping/MetadataCollection.php:
```diff
@@ -0,0 +1,47 @@
+     */
+    private $metadata;
+
+    private function __construct(array $metadata)
```
Document this please

Ocramius commented on lib/Doctrine/ORM/Mapping/MetadataCollection.php:
```diff
@@ -0,0 +1,47 @@
+        $this->metadata = $metadataByClassName;
+    }
+
+    public function get(string $name) : ClassMetadata
```
The various APIs getting class metadata by object reference could need a `getByObject(object $object)`

Ocramius commented on tests/Doctrine/Tests/OrmFunctionalTestCase.php:
```diff
@@ -743,8 +744,66 @@ protected function getEntityManager(?Connection $connection = null, ?MappingDriv
 
         $config->setMetadataDriverImpl(
             $mappingDriver ?? $config->newDefaultAnnotationDriver([
+                realpath(__DIR__ . '/Models/CMS'),
```
Wondering why this test changed to load everything - is this because we don't yet have a lazy collection?

mikeSimonson replied:
Just because I am pocking around a bit everywhere to try and fix everything that I broke.

Ocramius reviewed (changes requested):
> @Ocramius My main question now is to know if passing the whole mapping collection to the BasicEntityPersister is the way to go ? (commit extract the MetadataFactory from the basic entity persister and the following one)

@mikeSimonson I'd suggest that any component requiring metadata should just be given the collection. The `MetadataCollection` is now to be considered a "dumb" immutable structure, so it should be fine to pass it around.

Obviously, if only one entity metadata is required in a persister, then it should be passed in explicitly, but I don't know how big that change would be.

It will be easy to extract these scenarios when looking at classes that always `get()` on the collection with the same parameter.

Ocramius commented:
@mikeSimonson I've updated the PR description with the TODOs matching the discussion above.

Majkl578 commented:
@mikeSimonson Do you plan to continue with this?

mikeSimonson commented:
@Majkl578 yes. I don't have much time right now but I hope that I will have it in a good state for amsterdamphp

derrabus commented:
I'm closing this PR because it hasn't seen progress for several years. Please open a new one if you want to pick up the work one day.
